galleon-plugins/src/main/java/org/wildfly/galleon/plugin/server/ForkedEmbeddedUtil.java

```diff
@@ -117,32 +125,38 @@ public static Path storeSystemProps() throws ProvisioningException {
         return props;
     }
 
-    public static void main(String... args) throws ProvisioningException {
-
-        if(args.length < 2) {
-            throw new IllegalStateException("Expected at least two arguments but got " + Arrays.asList(args));
-        }
+    public static void main(String... args) {
+        try{
+            if(args.length < 2) {
+                throw new IllegalStateException("Expected at least two arguments but got " + Arrays.asList(args));
+            }
 
-        // set system properties
-        setSystemProps(args[0]);
+            // set system properties
+            setSystemProps(args[0]);
 
-        Class<?> cls;
-        try {
-            cls = Thread.currentThread().getContextClassLoader().loadClass(args[1]);
-        } catch (ClassNotFoundException e) {
-            throw new ProvisioningException("Failed to locate the target class " + args[1], e);
-        }
+            Class<?> cls;
+            try {
+                cls = Thread.currentThread().getContextClassLoader().loadClass(args[1]);
+            } catch (ClassNotFoundException e) {
+                throw new ProvisioningException("Failed to locate the target class " + args[1], e);
+            }
         /*if(!ForkCallback.class.isAssignableFrom(cls)) {
             throw new ProvisioningException(args[1] + " does not implement " + ForkCallback.class.getName());
         }*/
 
-        Object o;
-        try {
-            o = cls.newInstance();
-        } catch (Exception e) {
-            throw new ProvisioningException("Failed to instantiate " + args[1], e);
+            Object o;
+            try {
+                o = cls.newInstance();
+            } catch (Exception e) {
+                throw new ProvisioningException("Failed to instantiate " + args[1], e);
+            }
+            ((ForkCallback)o).forkedForEmbedded(args.length == 2 ? new String[0] : Arrays.copyOfRange(args, 2, args.length));
+
+            System.exit(NORMAL);
+        } catch (Throwable t){
+            t.printStackTrace(System.err);
+            System.exit(FAILED);
```

Comment:
Let's avoid System.exit(...) calls. Simply re-throw the throwable from the catch.

Reply:
@aloubyansky I would say re-throw the exception was the root of the hang. As far as I can understand, the caller doesn't receive the signal about the process has finished if we throw an exception from the main method. Basically, the caller hangs reading the input stream.